### PR TITLE
Fix handling wkt strings and add pyproj.CRS support

### PR DESCRIPTION
**Pull request by ahuang11**
Closes https://github.com/holoviz/hvplot/issues/1137

The issue was that [rioxarray gets loaded after running the second cell](https://github.com/holoviz/hvplot/issues/1137#issuecomment-1719891849), which triggered the first if clause, but we didn't handle WKT strings properly; this PR fixes that.
```python
    def _process_crs(self, data, crs):
        """Given crs as proj4 string, data.attr, or cartopy.crs return cartopy.crs
        """
        if hasattr(data, 'rio') and data.rio.crs is not None:
            # if data is a rioxarray
            _crs = data.rio.crs.to_wkt()
            print("RIO")
        else:
            # get the proj string: either the value of data.attrs[crs] or crs itself
            _crs = getattr(data, 'attrs', {}).get(crs or 'crs', crs)
            print("NOT RIO")
```

Supersedes contaminated https://github.com/holoviz/hvplot/pull/1138 

Note the notebook cell execution number
image

### CODE REVIEW

**Comment by Hoxbro** [comment]
Thank you for finding the cause of the confusing behavior. Your changes stops the error, but I still find the behavior not to be ideal, e.g., 

``` python
import cartopy.crs as ccrs
import rioxarray
import xarray as xr
from hvplot.util import process_crs

crs = ccrs.RotatedPole(pole_longitude=177.5, pole_latitude=37.5)
ds = xr.Dataset(attrs={"crs": crs})

a1 = process_crs(ds.rio.crs.to_proj4())
a2 = process_crs(ds.attrs["crs"])

type(a1), type(a2)
```

Outputs: `(cartopy.crs.PlateCarree, cartopy.crs.RotatedPole)`

I think we should swap the order of the if statement.

**Comment by ahuang11** [comment]
Good catch! I believe `if proj_is_latlong` is making a wrong assumption that only PlateCarree is lat lon; rotated pole is also in latlong.
```
    A rotated latitude/longitude projected coordinate system
    with cylindrical topology and projected distance.
```

File: hvplot/util.py [hunk 1]
```diff
@@ -213,6 +217,8 @@ def proj_to_cartopy(proj):
     else:
         kw_proj.pop('latitude_true_scale', None)
 
+    print(kw_proj)
+
```
Comment (ahuang11) [inline hvplot/util.py:221]:
```suggestion

```

File: hvplot/util.py [hunk 2]
```diff
@@ -252,13 +260,12 @@ def process_crs(crs):
...
         try:
-            crs = crs[5:].lstrip().rstrip()
-            return ccrs.epsg(crs)
+            crs = pyproj.CRS.from_epsg(crs).to_wkt()
         except Exception as e:
             errors.append(e)
     if isinstance(crs, int):
         try:
-            return ccrs.epsg(crs)
+            crs = pyproj.CRS.from_epsg(crs).to_wkt()
```
Comment (Hoxbro) [inline hvplot/util.py:268]: Is this really needed? I would expect an integer to either fail or pass without converting it to_wkt.

File: hvplot/util.py [hunk 3]
```diff
@@ -167,19 +170,23 @@ def proj_to_cartopy(proj):
...
+            elif v == 'utm':
                 cl = ccrs.UTM
-            if v == 'stere':
+            elif v == 'stere':
                 cl = ccrs.Stereographic
-            if v == 'ob_tran':
+            elif v == 'ob_tran':
                 cl = ccrs.RotatedPole
+            else:
+                raise NotImplementedError('Unknown projection {}'.format(v))
```
Comment (Hoxbro) [inline hvplot/util.py:189]: nit: could be f-string

File: hvplot/util.py [hunk 4]
```diff
@@ -252,13 +260,12 @@ def process_crs(crs):
     errors = []
     if isinstance(crs, str) and crs.lower().startswith('epsg'):
         try:
-            crs = crs[5:].lstrip().rstrip()
-            return ccrs.epsg(crs)
+            crs = pyproj.CRS.from_epsg(crs).to_wkt()
```
Comment (Hoxbro) [inline hvplot/util.py:263]: Again, is this needed? There is no need to do extra calculations if the simple case works.

File: hvplot/util.py (hunk 2 above)
Comment (ahuang11) [reply hvplot/util.py:268]:
```
import cartopy.crs as ccrs

ccrs.epsg(4326)
```

Comment (ahuang11) [reply hvplot/util.py:268]:
```
import pyproj
import cartopy.crs as ccrs
from hvplot.util import process_crs

process_crs(pyproj.CRS.from_epsg(4326).to_wkt())
```

File: hvplot/util.py (hunk 3 above)
Comment (ahuang11) [inline hvplot/util.py:189]:
```suggestion
                raise NotImplementedError(f'Unknown projection {v}')
```

File: hvplot/util.py (hunk 4 above)
Comment (ahuang11) [reply hvplot/util.py:263]: The simple case unfortunately does not work

File: hvplot/util.py (hunk 2 above)
Comment (Hoxbro) [reply hvplot/util.py:268]: The logic around it is set up so it is okay for a function to fail. So the question is, does `process_crs(4326)` work both before and after this PR?

Comment (ahuang11) [reply hvplot/util.py:268]: Before, yes. However, due to this being removed for having incorrect assumptions, no.
```
    if proj_is_latlong(proj):
        return ccrs.PlateCarree()
```

Also to add, ccrs.epsg(3857) vs ccrs.epsg("3857") on main yields different results, which can be bad.
image

Comment (ahuang11) [reply hvplot/util.py:268]: This branch unifies the string/int input:

image

Comment (Hoxbro) [reply hvplot/util.py:268]: Even though the projections are different types, can't they represent the same projection?

image

https://epsg.io/3857

If you feel strongly about always using `pyproj.CRS.from_epsg(xxx).to_wkt()`. I don't have any objection, but think we could combine some of the if statements.

Comment (ahuang11) [reply hvplot/util.py:268]: Okay, I'll try to simplify the if statements.

File: hvplot/util.py [hunk 5]
```diff
@@ -62,7 +62,7 @@ def check_crs(crs):
 
     Returns
     -------
-    A valid crs if possible, otherwise None
+    A valid crs if possible, otherwise raises an error.
```
Comment (ahuang11) [inline hvplot/util.py:65]:
```suggestion
    A valid crs if possible, otherwise None.
```

File: hvplot/tests/testutil.py [hunk 6]
```diff
@@ -283,20 +291,58 @@ def test_check_crs():
...
+
+
+@pytest.mark.parametrize("input", [
+    "3857",
+    3857,
+    "epsg:3857",
+    "EPSG: 3857",
+    "+init=epsg:3857",
     'PROJCS["WGS 84 / Pseudo-Mercator",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4326"]],PROJECTION["Mercator_1SP"],PARAMETER["central_meridian",0],PARAMETER["scale_factor",1],PARAMETER["false_easting",0],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH],EXTENSION["PROJ4","+proj=merc +a=6378137 +b=6378137 +lat_ts=0 +lon_0=0 +x_0=0 +y_0=0 +k=1 +units=m +nadgrids=@null +wktext +no_defs"],AUTHORITY["EPSG","3857"]]',
-    # Created with pyproj.CRS("EPSG:3857").to_wkt()
```
Comment (Hoxbro) [inline hvplot/tests/testutil.py:291]: Could be nice to still have this comment.

File: hvplot/tests/testutil.py [hunk 7]
```diff
@@ -283,20 +291,58 @@ def test_check_crs():
 
 @pytest.mark.parametrize("input", [
     "+init=epsg:26911",
+], ids=lambda x: str(x)[:20])
```
Comment (Hoxbro) [inline hvplot/tests/testutil.py:294]:
```suggestion
])
```
NIT: The only reason I added ids, was not to have a big string in the CI for the long proj definition

**Review by Hoxbro — APPROVED** [review]
It looks good to me.

One small thing is we have identical functions in GeoViews. Can you update them and test that we don't break anything in that CI, too, with these changes? 

I think it is out-of-scope for this PR, but we are getting some warnings related to these tests. (Already there before this PR)

``` python
hvplot/tests/testgeo.py::TestProjections::test_plot_with_crs_as_nonexistent_attr_str
hvplot/tests/testutil.py::TestGeoUtil::test_proj_to_cartopy
hvplot/tests/testutil.py::test_check_crs
hvplot/tests/testutil.py::test_check_crs
hvplot/tests/testutil.py::test_process_crs[+init=epsg:26911]
hvplot/tests/testutil.py::test_process_crs_pyproj_proj
hvplot/tests/testutil.py::test_process_crs_platecarree[+init=epsg:4326]
hvplot/tests/testutil.py::test_process_crs_mercator[+init=epsg:3857]
  /Users/runner/miniconda3/envs/test-environment/lib/python3.11/site-packages/pyproj/crs/crs.py:141: FutureWarning: '+init=<authority>:<code>' syntax is deprecated. '<authority>:<code>' is the preferred initialization method. When making the change, be mindful of axis order changes: https://pyproj4.github.io/pyproj/stable/gotchas.html#axis-order-changes-in-proj-6
    in_crs_string = _prepare_from_proj_string(in_crs_string)
```

File: hvplot/tests/testutil.py [hunk 8]
```diff
@@ -283,20 +291,58 @@ def test_check_crs():
...
...semble",MEMBER["World Geodetic System 1984 (Transit)"],MEMBER["World Geodetic System 1984 (G730)"],MEMBER["World Geodetic System 1984 (G873)"],MEMBER["World Geodetic System 1984 (G1150)"],MEMBER["World Geodetic System 1984 (G1674)"],MEMBER["World Geodetic System 1984 (G1762)"],MEMBER["World Geodetic System 1984 (G2139)"],ELLIPSOID["WGS 84",6378137,298.257223563,LENGTHUNIT["metre",1]],ENSEMBLEACCURACY[2.0]],PRIMEM["Greenwich",0,ANGLEUNIT["degree",0.0174532925199433]],ID["EPSG",4326]],CONVERSION["Popular Visualisation Pseudo-Mercator",METHOD["Popular Visualisation Pseudo Mercator",ID["EPSG",1024]],PARAMETER["Latitude of natural origin",0,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8801]],PARAMETER["Longitude of natural origin",0,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8802]],PARAMETER["False easting",0,LENGTHUNIT["metre",1],ID["EPSG",8806]],PARAMETER["False northing",0,LENGTHUNIT["metre",1],ID["EPSG",8807]]],CS[Cartesian,2],AXIS["easting (X)",east,ORDER[1],LENGTHUNIT["metre",1]],AXIS["northing (Y)",north,ORDER[2],LENGTHUNIT["metre",1]],USAGE[SCOPE["Web mapping and visualisation."],AREA["World between 85.06°S and 85.06°N."],BBOX[-85.06,-180,85.06,180]],ID["EPSG",3857]]',
```
Comment (ahuang11) [inline hvplot/tests/testutil.py:338]:
```suggestion
    # Created with pyproj.CRS("EPSG:4326").to_wkt()
    'PROJCRS["WGS 84 / Pseudo-Mercator",BASEGEOGCRS["WGS 84",ENSEMBLE["World Geodetic System 1984 ensemble",MEMBER["World Geodetic System 1984 (Transit)"],MEMBER["World Geodetic System 1984 (G730)"],MEMBER["World Geodetic System 1984 (G873)"],MEMBER["World Geodetic System 1984 (G1150)"],MEMBER["World Geodetic System 1984 (G1674)"],MEMBER["World Geodetic System 1984 (G1762)"],MEMBER["World Geodetic System 1984 (G2139)"],ELLIPSOID["WGS 84",6378137,298.257223563,LENGTHUNIT["metre",1]],ENSEMBLEACCURACY[2.0]],PRIMEM["Greenwich",0,ANGLEUNIT["degree",0.0174532925199433]],ID["EPSG",4326]],CONVERSION["Popular Visualisation Pseudo-Mercator",METHOD["Popular Visualisation Pseudo Mercator",ID["EPSG",1024]],PARAMETER["Latitude of natural origin",0,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8801]],PARAMETER["Longitude of natural origin",0,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8802]],PARAMETER["False easting",0,LENGTHUNIT["metre",1],ID["EPSG",8806]],PARAMETER["False northing",0,LENGTHUNIT["metre",1],ID["EPSG",8807]]],CS[Cartesian,2],AXIS["easting (X)",east,ORDER[1],LENGTHUNIT["metre",1]],AXIS["northing (Y)",north,ORDER[2],LENGTHUNIT["metre",1]],USAGE[SCOPE["Web mapping and visualisation."],AREA["World between 85.06°S and 85.06°N."],BBOX[-85.06,-180,85.06,180]],ID["EPSG",3857]]',
```

File: hvplot/tests/testutil.py [hunk 9]
```diff
@@ -283,20 +291,59 @@ def test_check_crs():
...
+    # Created with pyproj.CRS("EPSG:4326").to_wkt()
```
Comment (ahuang11) [inline hvplot/tests/testutil.py:338]:
```suggestion
    # Created with pyproj.CRS("EPSG:3857").to_wkt()
```

File: hvplot/tests/testutil.py [hunk 10]
```diff
@@ -283,20 +291,59 @@ def test_check_crs():
...
+    assert isinstance(crs, ccrs.PlateCarree)
+
+
+@pytest.mark.parametrize("input", [
     "4326",
     4326,
     "epsg:4326",
     "EPSG: 4326",
+    "+init=epsg:4326",
+    'GEOGCRS["WGS 84",ENSEMBLE["World Geodetic System 1984 ensemble",MEMBER["World Geodetic System 1984 (Transit)"],MEMBER["World Geodetic System 1984 (G730)"],MEMBER["World Geodetic System 1984 (G873)"],MEMBER["World Geodetic System 1984 (G1150)"],MEMBER["World Geodetic System 1984 (G1674)"],MEMBER["World Geodetic System 1984 (G1762)"],MEMBER["World Geodetic System 1984 (G2139)"],ELLIPSOID["WGS 84",6378137,298.257223563,LENGTHUNIT["metre",1]],ENSEMBLEACCURACY[2.0]],PRIMEM["Greenwich",0,ANGLEUNIT["degree",0.0174532925199433]],CS[ellipsoidal,2],AXIS["geodetic latitude (Lat)",north,ORDER[1],ANGLEUNIT["degree",0.0174532925199433]],AXIS["geodetic longitude (Lon)",east,ORDER[2],ANGLEUNIT["degree",0.0174532925199433]],USAGE[SCOPE["Horizontal component of 3D system."],AREA["World."],BBOX[-90,-180,90,180]],ID["EPSG",4326]]',
```
Comment (ahuang11) [inline hvplot/tests/testutil.py:322]:
```suggestion
    # Created with pyproj.CRS("EPSG:4326").to_wkt()
    'GEOGCRS["WGS 84",ENSEMBLE["World Geodetic System 1984 ensemble",MEMBER["World Geodetic System 1984 (Transit)"],MEMBER["World Geodetic System 1984 (G730)"],MEMBER["World Geodetic System 1984 (G873)"],MEMBER["World Geodetic System 1984 (G1150)"],MEMBER["World Geodetic System 1984 (G1674)"],MEMBER["World Geodetic System 1984 (G1762)"],MEMBER["World Geodetic System 1984 (G2139)"],ELLIPSOID["WGS 84",6378137,298.257223563,LENGTHUNIT["metre",1]],ENSEMBLEACCURACY[2.0]],PRIMEM["Greenwich",0,ANGLEUNIT["degree",0.0174532925199433]],CS[ellipsoidal,2],AXIS["geodetic latitude (Lat)",north,ORDER[1],ANGLEUNIT["degree",0.0174532925199433]],AXIS["geodetic longitude (Lon)",east,ORDER[2],ANGLEUNIT["degree",0.0174532925199433]],USAGE[SCOPE["Horizontal component of 3D system."],AREA["World."],BBOX[-90,-180,90,180]],ID["EPSG",4326]]',
```

**Pull request merged**